### PR TITLE
BE-519: Set up OpenTelemetry across hash-api and Temporal workers

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Wires OpenTelemetry through hash-api, hash-ai-worker-ts, and hash-integration-worker so traces, logs, and metrics flow into the existing OTLP collector. Caller-side trace context (Express HTTP spans, Rust `temporal-client::start_workflow`) is propagated into Temporal workflow start headers so the worker-side `RunWorkflow` and `RunActivity` spans chain off the caller's trace in Tempo.

Outbound `fetch` calls (OpenAI, Anthropic, Linear, …) are now traced via `@opentelemetry/instrumentation-undici` with a shared `peer.service` mapping so Tempo's `service_graphs` processor renders external dependencies as edges in the service map.

## 🔗 Related links

- [BE-519](https://linear.app/hash/issue/BE-519)
- [BE-520](https://linear.app/hash/issue/BE-520) — follow-up for dropping the v1↔v2 SDK adapter once `@temporalio/interceptors-opentelemetry-v2` is released
- [SRE-676](https://linear.app/hash/issue/SRE-676) — follow-up for Temporal Server self-instrumentation

## 🚫 Blocked by

- [ ] none

## 🔍 What does this change?

**New shared OTEL module** (`@local/hash-backend-utils/opentelemetry`):

- `registerOpenTelemetry({ endpoint, serviceName, instrumentations })` registers a global `NodeTracerProvider` / `LoggerProvider` / `MeterProvider` against an OTLP/gRPC collector.
- `createUndiciInstrumentation()` and `httpRequestSpanNameHook` shared across all three services.
- `peer.service` mapping (discriminated exact/suffix host rules) for OpenAI, Anthropic, Linear, Google Cloud.
- `BatchSpanProcessor` / `BatchLogRecordProcessor` keep export off the request path. Shutdown surfaces `Promise.allSettled` rejections and applies a 2-second per-provider timeout.

**Temporal trace propagation:**

- Workflow client interceptor (`OpenTelemetryWorkflowClientInterceptor`) attached in `createTemporalClient` injects the active context into workflow start headers.
- Activity-side interceptors (`activity` factory shape) extract context and stamp `trace_id` / `span_id` on activity log lines for Loki ↔ Tempo correlation.
- Workflow modules registered both in `bundleWorkflowCode` (production path) AND in `Worker.create` (dev path) — without the bundle registration, prebuilt bundles ignore the runtime `workflowModules` config and prod ships zero workflow spans.
- `wrapWorkflowSpanExporter` / `makeV2WorkflowSink` bridges v1-shaped `ReadableSpan` from `@temporalio/interceptors-opentelemetry@1` to our v2 stack (synthesises `instrumentationScope` and `parentSpanContext`); marked `TODO(BE-520)` and quarantined to one helper.

**Rust trace context propagation** (`hash-temporal-client::ai`):

- Rewrites `start_ai_workflow` to use the low-level `WorkflowService::start_workflow_execution` so the proto `header` field is exposed.
- `build_otel_header()` injects the active trace context as a `_tracer-data` payload; emits a once-per-process `tracing::warn!` when the carrier is empty.
- Caller span annotated with `otel.kind = "producer"` for the asynchronous fire-and-forget shape.

**Worker bootstrap helper** (`@local/hash-backend-utils/temporal/worker-bootstrap`):

- `runWorker(opts)` collapses the previously-duplicated bootstrap logic in both worker `main.ts` files. `Sentry.init` stays per-worker because of ESM import-ordering.
- `WorkflowSource` discriminated union (`{ kind: "bundle" } | { kind: "path" }`) replaces `Partial<WorkerOptions>`; `ExtraWorkerOptions = Omit<WorkerOptions, …>` excludes helper-owned and source-owned keys from the escape hatch.
- SIGTERM/SIGINT handler awaits `worker.run()` so in-flight activities drain before OTEL providers shut down. Exit code 0 on graceful signals.

**Sentry coexistence:**

- All three services pass `skipOpenTelemetrySetup: !!otelSetup` so Sentry shares our `NodeTracerProvider`. If Sentry registered its own, it would shadow the global provider and the OTEL workflow client interceptor would silently break trace context propagation.

**Metrics scrape:**

- `apps/hash-external-services/opentelemetry-collector/otel-collector-config.yaml` now scrapes Temporal Server's prometheus endpoint (`temporal:8000`) with a `temporal_` prefix relabel.

### 🚢 Has this modified a publishable library?

This PR:

- [x] does not modify any publishable blocks or libraries, or modifications do not need publishing

### 📜 Does this require a change to the docs?

The changes in this PR:

- [x] are internal and do not require a docs change

### 🕸️ Does this require a change to the Turbo Graph?

The changes in this PR:

- [x] do not affect the execution graph

## ⚠️ Known issues

- The v1↔v2 SDK adapter is a workaround until `@temporalio/interceptors-opentelemetry-v2` lands — tracked in BE-520.
- Caller→workflow tracing uses parent-child semantics (what `@temporalio/interceptors-opentelemetry` produces). The OTEL spec for async messaging recommends PRODUCER/CONSUMER + Span Links, but neither v1 nor v2 of the upstream package implement that. Doesn't affect our metrics today: workflow spans are `INTERNAL` kind and start *after* the caller span ends, so Tempo's service-graphs processor doesn't inflate edge latencies. Captured as a follow-up note in BE-520.
- `HASH_OTLP_ENDPOINT` must be set on ALL backend services or NONE — a partial config silently breaks caller↔worker context propagation. No runtime check (deferred — infra-level concern).

## 🐾 Next steps

- BE-520: switch to `@temporalio/interceptors-opentelemetry-v2` once released, drop the workflow span adapter.
- SRE-676: instrument Temporal Server itself.
- BE-518: broader logger cleanup sweep (separate PR).

## 🛡 What tests cover this?

23 unit tests in `libs/@local/hash-backend-utils/`:

- `src/opentelemetry.test.ts` (12 tests) — `resolvePeerService` (exact/suffix matching, lookalike-domain rejection, suffix-dot-boundary), `httpRequestSpanNameHook` (incoming, outgoing, Express-wrapped, query-stripped, missing-method paths).
- `src/temporal/workflow-span-adapter.test.ts` (11 tests) — v1→v2 normalisation (instrumentationLibrary → instrumentationScope, parentSpanId → parentSpanContext), already-v2 passthrough identity check, attribute / event preservation through the rewrite path, mixed batch handling, the `parentSpanId === ""` edge case, existing-`parentSpanContext` precedence, result-callback propagation, shutdown / forceFlush delegation.

Drain semantics on SIGTERM are verified manually — `TestWorkflowEnvironment` integration tests deferred.

## ❓ How to test this?

1. Set `HASH_OTLP_ENDPOINT=http://localhost:4317` and start the OTEL collector + Tempo + Mimir + Loki via the `apps/hash-external-services` docker-compose.
2. `yarn dev:backend` + `yarn start:worker:ai`.
3. From the frontend, create an entity (triggers the `updateEntityEmbeddings` workflow).
4. Grafana → Tempo → search by service `Node API`. Expected trace shape:
   - `Node API POST /graphql` (~150ms) at the root.
   - `start_ai_workflow` (~12ms) as child, marked `Producer` kind.
   - `AI Worker RunWorkflow:updateEntityEmbeddings` (~2.5s) as child of the producer.
   - `RunActivity:createAndStoreEntityEmbeddingsActivity` nested under it.
   - `OpenAI POST /v1/embeddings` as external-service span (undici instrumentation).
5. Service map should show nodes for `Node API`, `Graph API`, `AI Worker`, `OpenAI`, `Postgres`, with an `AI Worker → OpenAI` edge.
6. Drain check: `kill -TERM <ai-worker-pid>` while a workflow is mid-flight. Logs show `Received SIGTERM, exiting…`, then activity completion logs, then OTEL flush, then exit 0.

## 📹 Demo

_Tempo screenshots from local testing — see Linear BE-519 for image attachments._

### CODE REVIEW

**Review by CiaranMn — DISMISSED** [review]

File: libs/@local/hash-backend-utils/src/opentelemetry.ts [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+          try {
+            await shutdownWithTimeout(label, run);
+          } catch (error) {
+            // eslint-disable-next-line no-console
+            console.error("OpenTelemetry %s shutdown failed:", label, error);
+            throw error;
+          }
+        }),
+      );
+      unregisterInstrumentations();
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply libs/@local/hash-backend-utils/src/opentelemetry.ts:336]: Well, we would have lost the data at this point anyway. But it also doesn't hurt to have a non-zero exit code I guess so it's visible in AWS.

**Review by CiaranMn — DISMISSED** [review]

File: libs/@local/hash-backend-utils/src/temporal/worker-bootstrap.ts [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+        logger.error("Worker shutdown trigger failed", { error });
+      }
+    }
+    // If the worker hasn't been created yet, do nothing more here:
+    // `runWorker` is still in startup and the early-exit branch below
+    // will fall through to the linear cleanup path when control
+    // returns from whichever `await` is pending.
+  };
+  process.on("SIGINT", onSignal);
+  process.on("SIGTERM", onSignal);
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply libs/@local/hash-backend-utils/src/temporal/worker-bootstrap.ts:213]: I won't fix this. This is pedantic.

**Review by vilkinsons — APPROVED** [review]
